test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -637,24 +639,30 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
 				b.Fatalf("op %d: %v", opIndex, err)
 			}
 			b.Cleanup(func() {
-				nsPreparer.cleanup()
+				if err := nsPreparer.cleanup(); err != nil {
+					b.Fatalf("failed to cleanup namespace: %v", err)
+				}
 			})
 
 		case *createPodsOp:
 			var namespace string
 			if concreteOp.Namespace != nil {
 				namespace = *concreteOp.Namespace
 			} else {
-				// define Pod's namespace automatically, and create that namespace.
+				// define Pod's namespace automatically.
 				namespace = fmt.Sprintf("namespace-%d", opIndex)
-				_, err := client.CoreV1().Namespaces().Create(ctx, &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}}, metav1.CreateOptions{})
-				if err != nil && !apierrors.IsAlreadyExists(err) {
+			}
+			// create namespace in case it's not created yet.
+			_, err := client.CoreV1().Namespaces().Create(ctx, &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}}, metav1.CreateOptions{})
```

Comment:
I don't think this will slow down the test that much because the time to create a namespace is trivial compared to creating 5000 pods, etc. This change creates one namespace per one `createPods`, not creates 5000 namespaces for 5000 pods created.
So, for example, this change only creates two namespaces in `SchedulingBasic/5000Nodes` test.
https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler_perf/config/performance-config.yaml#L18

What do you think? (Sorry if you've already figured these things out and are giving advice. 🙏  )

Reply:
https://github.com/kubernetes/kubernetes/pull/107771#discussion_r792342599

As I said upper. I also think it better to `createNamespaces` before `createPods` in the `workloadTemlate` other than create namespace in `createPods`.  This will help us to separate the process into different single op. Just personal advice. Thanks

Reply:
Yeah, that's another design choice we can choose; let `createPods` create only Pod.
But, currently, scheduler_perf is designed in such a way that `createPods` makes related resources. We also create resources other than namespace in `createPods` like PV, PVC. 
So, personally, it doesn't seem strange to me that `createPods` create namespaces. But, no strong opinion here.

What do you think?

Reply:
> I don't think this will slow down the test that much because the time to create one namespace is trivial compared to creating 500~10000 pods. 

Even if it only *creates* one namespace for all pods, it *tries* for every single one of them, which causes additional load. One can argue that it is the job of a benchmark to generate load and see how the system handles it, but this approach here is unusual and thus not representative for real deployments.

How about tracking in the function which namespaces are known to exist? The prior to the `Create` call check that and skip the call when it isn't necessary anymore.

Reply:
> How about tracking in the function which namespaces are known to exist? The prior to the Create call check that and skip the call when it isn't necessary anymore.

I am not against this. But,  if `skipWaitToCompletion` isn't `true`, namespaces should be unique to check if all newly created Pods are scheduled [here](https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler_perf/scheduler_perf_test.go#L897) by fetching all Pods in namespace, so the namespace creation will not be(should not be) skipped in most cases.

---

IIUC, the only reason we're creating namespace in each `createPods` is to check if all newly created Pods are scheduled. So, how about this idea:
- create one namespace when start scheduler_perf and use that namespace if `createPodsOp.Namespace == nil` instead of creating new namespace for each `createPods`.
  - if `createPodsOp.Namespace != nil`, the namespace won't be created.
- Pods created in the same `createPods` will have the same label.
- check if all newly created Pods are scheduled by fetching with label selector, instead of fetching all Pods in a namespace.

This idea seems to avoid all of the concerns raised in this thread.

Reply:
> Even if it only creates one namespace for all pods, it tries for every single one of them, which causes additional load.

Exactly. It's overwhelming the API server, using QPS, etc.

I think your proposal is fine. I think some of the tests require multiple namespaces (like the ones for pod affinity). But they should be creating and setting the desired namespace instead of getting it implicitly.